### DISCUSSION TITLE
What is the best practice to define a function that will be used in different script?

### DISCUSSION THREAD

**Discussion by Oyyko** [xmake-io/xmake#4070] [Q&A] [answered]
For example: I have to target A and B. And I want to define a function that will be used in the `after_install()` script in two targets. What is the best way to get this done?

Now, I just create a file `./xmake_modules/global.lua` and define a module in this file. Then in the `after_install()` function of `xmake.lua`, I just use 

```lua
import("xmake_modules.global")
global.my_functions(target)
```

This way really works for me. I just want to know if this is a good way to implement so-called "global function&variables" in the script scope of xmake.lua. Thanks!

In #3926 , I found something very useful. But I think the background of #3926 is about to define variables. So the first and second way in https://github.com/xmake-io/xmake/discussions/3926#discussioncomment-6357866 is not properate for a function definition.  The third way in https://github.com/xmake-io/xmake/discussions/3926#discussioncomment-6357866 is just my way (use a new script file). 

No matter what, I believe my issue has been resolved in #3926. However, I think bringing up this problem would be helpful for future users conducting searches. Therefore, I still intend to post this question.

**Comment by waruqi** [comment, marked as answer]
you can define function in a single file.

or use main entry.

```lua
after_install("xmake_modules.global")
```

./xmake_modules/global.lua

```lua
function main()
end
```

you can also use `add_moduledirs` to add modules root directory.
 
```lua
add_moduledirs("xmake_modules")
target("test")
    after_install("global")
```
